rust-sdk/core/src/quote/fees.rs

```diff
@@ -0,0 +1,210 @@
+use ethnum::U256;
+
+#[cfg(feature = "wasm")]
+use orca_whirlpools_macros::wasm_expose;
+
+use crate::{
+    try_apply_transfer_fee, CollectFeesQuote, ErrorCode, PositionFacade, TickFacade, TransferFee,
+    WhirlpoolFacade, AMOUNT_EXCEEDS_MAX_U64, ARITHMETIC_OVERFLOW,
+};
+
+/// Calculate fees owed for a position
+///
+/// # Paramters
+/// - `whirlpool`: The whirlpool state
+/// - `position`: The position state
+/// - `tick_lower`: The lower tick state
+/// - `tick_upper`: The upper tick state
+/// - `transfer_fee_a`: The transfer fee for token A
+/// - `transfer_fee_b`: The transfer fee for token B
+///
+/// # Returns
+/// - `CollectFeesQuote`: The fees owed for token A and token B
+#[allow(clippy::too_many_arguments)]
+#[cfg_attr(feature = "wasm", wasm_expose)]
+pub fn collect_fees_quote(
+    whirlpool: WhirlpoolFacade,
+    position: PositionFacade,
+    tick_lower: TickFacade,
+    tick_upper: TickFacade,
+    transfer_fee_a: Option<TransferFee>,
+    transfer_fee_b: Option<TransferFee>,
+) -> Result<CollectFeesQuote, ErrorCode> {
+    let mut fee_growth_below_a: u128 = tick_lower.fee_growth_outside_a;
+    let mut fee_growth_above_a: u128 = tick_upper.fee_growth_outside_a;
+    let mut fee_growth_below_b: u128 = tick_lower.fee_growth_outside_b;
+    let mut fee_growth_above_b: u128 = tick_upper.fee_growth_outside_b;
+
+    if whirlpool.tick_current_index < position.tick_lower_index {
+        fee_growth_below_a = whirlpool.fee_growth_global_a - fee_growth_below_a;
+        fee_growth_below_b = whirlpool.fee_growth_global_b - fee_growth_below_b;
+    }
+
+    if whirlpool.tick_current_index >= position.tick_upper_index {
+        fee_growth_above_a = whirlpool.fee_growth_global_a - fee_growth_above_a;
+        fee_growth_above_b = whirlpool.fee_growth_global_b - fee_growth_above_b;
+    }
+
+    let fee_growth_inside_a =
+        whirlpool.fee_growth_global_a - fee_growth_below_a - fee_growth_above_a;
+
+    let fee_growth_inside_b =
+        whirlpool.fee_growth_global_b - fee_growth_below_b - fee_growth_above_b;
+
+    let fee_owed_delta_a: U256 =
+        <U256>::from(fee_growth_inside_a - position.fee_growth_checkpoint_a)
+            .checked_mul(position.liquidity.into())
+            .ok_or(ARITHMETIC_OVERFLOW)?
+            >> 64;
+
+    let fee_owed_delta_b: U256 =
+        <U256>::from(fee_growth_inside_b - position.fee_growth_checkpoint_b)
+            .checked_mul(position.liquidity.into())
+            .ok_or(ARITHMETIC_OVERFLOW)?
+            >> 64;
+
```

Comment:
Isn't this the same as a u128 sub underflow?

All of `fee_growth_inside_a`, `position.fee_growth_checkpoint_a`, `fee_growth_inside_b`, and `position.fee_growth_checkpoint_b` are u128. IIRC in rust the default operators (`-`, `+`, `*`, etc.) will panic in debug mode but happily overflow (and wrap around) in release mode (https://doc.rust-lang.org/rustc/codegen-options/index.html#overflow-checks)

Reply:
Thanks for letting me know this behavior.
TBH, I don't know this behavior.

The reason why the SDK uses subUnderflowU128 and the contract uses wrapping_sub is because it ‘intentionally’ manages values cyclically, with or without overflow checks.

The use of wrapping_sub tells the reader that this code explicitly allows for underflow. It is not clear whether a mere subtraction is simply an omission of consideration or whether there is a deeper meaning.

The result may be the same in an environment without overflow checking in release mode, but the code in the contract should be copied. As mentioned above, clear declarative information that underflow is not a problem is lost.

Also, wouldn't it result in a non-working SDK if the developer tested it in debug mode?